tools/draincluster/drain/drain.go

```diff
@@ -0,0 +1,212 @@
+		if validCondition != nil && validCondition.Status == metav1.ConditionFalse {
+			// check to see if CRP is missing or CRP is being deleted or CRB is missing.
+			if validCondition.Reason == condition.EvictionInvalidMissingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidDeletingCRPMessage ||
+				validCondition.Reason == condition.EvictionInvalidMissingCRBMessage {
+				continue
+			}
+		}
+		executedCondition := eviction.GetCondition(string(placementv1beta1.PlacementEvictionConditionTypeExecuted))
+		if executedCondition == nil || executedCondition.Status == metav1.ConditionFalse {
+			isDrainSuccessful = false
+			log.Printf("eviction %s was not executed successfully for CRP %s", evictionName, crpName)
+			continue
+		}
+		log.Printf("eviction %s was executed successfully for CRP %s", evictionName, crpName)
+		// log each cluster scoped resource evicted for CRP.
+		clusterScopedResourceIdentifiers, err := h.collectClusterScopedResourcesSelectedByCRP(ctx, crpName)
+		if err != nil {
+			log.Printf("failed to collect cluster scoped resources selected by CRP %s: %v", crpName, err)
+			continue
+		}
+		for _, resourceIdentifier := range clusterScopedResourceIdentifiers {
+			log.Printf("evicted resource %s propagated by CRP %s", generateResourceIdentifierKey(resourceIdentifier), crpName)
+		}
+	}
+
+	return isDrainSuccessful, nil
+}
+
+func (h *Helper) cordon(ctx context.Context) error {
+	// add taint to member cluster to ensure resources aren't scheduled on it.
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		var mc clusterv1beta1.MemberCluster
+		if err := h.HubClient.Get(ctx, types.NamespacedName{Name: h.ClusterName}, &mc); err != nil {
+			return err
+		}
+
+		// search to see cordonTaint already exists on the cluster.
+		for i := range mc.Spec.Taints {
+			if mc.Spec.Taints[i] == toolsutils.CordonTaint {
+				return nil
+			}
+		}
+
+		// add taint to member cluster to cordon.
+		mc.Spec.Taints = append(mc.Spec.Taints, toolsutils.CordonTaint)
+
+		return h.HubClient.Update(ctx, &mc)
+	})
+}
+
+func (h *Helper) fetchClusterResourcePlacementNamesToEvict(ctx context.Context) ([]string, error) {
+	var crbList placementv1beta1.ClusterResourceBindingList
+	if err := h.HubClient.List(ctx, &crbList); err != nil {
+		return []string{}, fmt.Errorf("failed to list cluster resource bindings: %w", err)
+	}
+
+	crpNameMap := make(map[string]bool)
+	// find all unique CRP names for which eviction needs to occur.
+	for i := range crbList.Items {
+		crb := crbList.Items[i]
+		if crb.Spec.TargetCluster == h.ClusterName && crb.DeletionTimestamp == nil {
+			crpName, ok := crb.GetLabels()[placementv1beta1.CRPTrackingLabel]
+			if !ok {
+				return []string{}, fmt.Errorf("failed to get CRP name from binding %s", crb.Name)
+			}
+			if !crpNameMap[crpName] {
+				crpNameMap[crpName] = true
+			}
+		}
+	}
+
+	crpNames := make([]string, len(crpNameMap))
+	i := 0
+	for crpName := range crpNameMap {
```

Comment:
A nit: this cast from map to string array seems redundant (in the sense that the map itself can be used directly anyway); we could also use apimachinery Set if necessary.